src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/IndexStateManagementHistory.kt

```diff
@@ -0,0 +1,200 @@
+            scheduledRollover = threadPool.scheduleWithFixedDelay({ rolloverAndDeleteHistoryIndex() }, historyRolloverCheckPeriod, executorName())
+        }
+    }
+
+    private fun rolloverAndDeleteHistoryIndex() {
+        if (historyEnabled) rolloverHistoryIndex()
+        deleteOldHistoryIndex()
+    }
+
+    private fun rolloverHistoryIndex(): Boolean {
+        if (!indexStateManagementIndices.indexStateManagementIndexHistoryExists()) {
+            return false
+        }
+
+        // We have to pass null for newIndexName in order to get Elastic to increment the index count.
+        val request = RolloverRequest(IndexStateManagementIndices.HISTORY_WRITE_INDEX_ALIAS, null)
+        request.createIndexRequest.index(IndexStateManagementIndices.HISTORY_INDEX_PATTERN)
+            .mapping(_DOC, indexStateManagementIndices.indexStateManagementHistoryMappings, XContentType.JSON)
+        request.addMaxIndexDocsCondition(historyMaxDocs)
+        request.addMaxIndexAgeCondition(historyMaxAge)
+        val response = client.admin().indices().rolloversIndex(request).actionGet()
+        if (!response.isRolledOver) {
+            logger.info("${IndexStateManagementIndices.HISTORY_WRITE_INDEX_ALIAS} not rolled over. Conditions were: ${response.conditionStatus}")
+        }
+        return response.isRolledOver
+    }
+
+    @Suppress("SpreadOperator")
+    private fun deleteOldHistoryIndex() {
+        val indexToDelete = mutableListOf<String>()
+
+        val clusterStateRequest = ClusterStateRequest()
+        clusterStateRequest.clear().indices(IndexStateManagementIndices.HISTORY_ALL).metaData(true)
+        clusterStateRequest.local(true)
+        val strictExpandIndicesOptions = IndicesOptions.strictExpand()
+        clusterStateRequest.indicesOptions(strictExpandIndicesOptions)
+
+        val clusterStateResponse = client.admin().cluster().state(clusterStateRequest).actionGet()
+
+        for (entry in clusterStateResponse.state.metaData.indices()) {
+            val indexMetaData = entry.value
+            val creationTime = indexMetaData.creationDate
+
+            logger.info("CreationTime $creationTime $historyRetentionPeriod.millis}")
+            if ((Instant.now().toEpochMilli() - creationTime) > historyRetentionPeriod.millis) {
+                val alias = indexMetaData.aliases.firstOrNull { IndexStateManagementIndices.HISTORY_WRITE_INDEX_ALIAS == it.value.alias }
+                if (alias != null && historyEnabled) {
+                    // If index has write alias and history is enable, don't delete the index.
+                    break
+                }
+
+                indexToDelete.add(indexMetaData.index.name)
+            }
+        }
+
+        if (indexToDelete.isNotEmpty()) {
+            val deleteRequest = DeleteIndexRequest(*indexToDelete.toTypedArray())
+            val deleteResponse = client.admin().indices().delete(deleteRequest).actionGet()
+            if (!deleteResponse.isAcknowledged) {
```

Comment:
Is this a partial failure or if one index has an issue they all fail?

Reply:
I had similar question but my investigation was inconclusive because lack of documentation and hard to reproduce `acknowledge false` locally.

For now to be safe, I implemented back up delete one-by-one request and log any failure at that time.

Reply:
Seems like it will only happen if one of the indices is in a [current in prog snapshot](https://github.com/elastic/elasticsearch/blob/a5d9939ac119d7f836408341a562e7bbeeac1e98/server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataDeleteIndexService.java#L99)